src/relay/pass/to_anf.cc

```diff
@@ -0,0 +1,474 @@
+/*!
+ * Copyright (c) 2018 by Contributors
+ *
+ * \file to_anf.cc
+ *
+ * \brief Turn implicit sharing into observable sharing.
+ */
+#include <tvm/relay/pass.h>
+#include <tvm/relay/expr_functor.h>
+#include "let_list.h"
+#include "../../common/arena.h"
+
+namespace tvm {
+namespace relay {
+
+using common::LinkNode;
+using common::LinkedList;
+
+/* DependencyGraph track where an expr is used, in the output node.
+ * It allow us to traverse the graph in reverse order.
+ */
+class DependencyGraph {
+ public:
+  struct Node;
+  /*! \brief A node in the graph. */
+  struct Node {
+    LinkedList<Node*> output;
+  };
+  /*! \brief The node map that maps node to graph */
+  std::unordered_map<Expr, Node*, NodeHash, NodeEqual> node_map;
+  std::unordered_map<Node*, Expr> node_unmap;
```

Comment:
This is quite confusing, i understand you want to map node to Expr, you can use weak ref ptr in the node instead, see https://github.com/dmlc/tvm/blob/master/src/relay/pass/fuse_ops.cc#L82